### PR TITLE
feat: no empty blocks exceptions

### LINKED ISSUE #149: `no-empty-blocks` exceptions
add configurable exceptions for the `no-empty-blocks` in the following scenarios
- [ ] modifiers with only `_` in their body: useful for specific styles of Branching Tree Technique tests
- [ ] try-catch blocks

### PR DESCRIPTION
**Pull request by jahabeebs**
closes #149 

add configurable exceptions for the no-empty-blocks in the following scenarios

- [X] modifiers with only _ in their body: useful for specific styles of Branching Tree Technique tests
- [X]  try-catch blocks

### CODE REVIEW

File: lib/rules/best-practises/no-empty-blocks.js [hunk 1]
```diff
@@ -16,32 +20,53 @@ const meta = {
...
+          code: 'modifier onlyOwner { _; }',
+        },
       ],
       good: [
         { description: 'empty receive function', code: 'receive () external {}' },
         { description: 'empty fallback function', code: 'fallback () external {}' },
         {
-          description: 'empty constructor with member initialization list',
-          code: 'constructor(uint param) Foo(param) Bar(param*2) {}',
+          description: 'empty constructor with parent init',
```
Comment (juanpcapurro) [inline lib/rules/best-practises/no-empty-blocks.js:32]: why did you choose this other wording? if we want to be consistent with the solidity docs... there's not a clear name for this, as there is for C++ 
They seem to refer to it as  ...[modifier-style base constructor invocation](https://docs.soliditylang.org/en/latest/contracts.html#arguments-for-base-constructors) ?

File: lib/rules/best-practises/no-empty-blocks.js [hunk 2]
```diff
@@ -16,32 +20,53 @@ const meta = {
...
+          code: 'constructor(uint x) Base(x) {}',
         },
+        { description: 'empty modifier with _ allowed', code: 'modifier onlyOwner { _; }' },
+        { description: 'empty try-catch block allowed', code: 'try foo() {} catch {}' },
       ],
     },
     notes: [
       {
-        note: 'The rule ignores an empty constructor by default as long as parent contracts are being initialized. See "Empty Constructor" example.',
+        note: 'Empty constructor is ignored if the constructor has parent-initialization modifiers.',
```
Comment (juanpcapurro) [inline lib/rules/best-practises/no-empty-blocks.js:41]: I stand behind this way to describe it :cowboy_hat_face:

File: lib/rules/best-practises/no-empty-blocks.js [hunk 3]
```diff
@@ -50,27 +75,37 @@ class NoEmptyBlocksChecker extends BaseChecker {
   }
 
   Block(node) {
-    // ignore empty constructor body when it's being defined to call a parent
-    // constructor
-    if (node.parent.isConstructor && node.parent.modifiers.length > 0) {
+    // Skip constructor with modifiers
+    if (node.parent.isConstructor && node.parent.modifiers && node.parent.modifiers.length > 0) {
```
Comment (juanpcapurro) [inline lib/rules/best-practises/no-empty-blocks.js:79]: was this causing an early return due to an exception?

File: lib/rules/best-practises/no-empty-blocks.js [hunk 4]
```diff
@@ -16,32 +20,53 @@ const meta = {
...
+      allowEmptyCatch: { type: 'boolean' },
+    },
+  },
 }
 
 class NoEmptyBlocksChecker extends BaseChecker {
   constructor(reporter) {
     super(reporter, ruleId, meta)
+
+    const [, options = {}] = Array.isArray(this.reporter?.config?.[ruleId])
```
Comment (juanpcapurro) [inline lib/rules/best-practises/no-empty-blocks.js:64]: what was the reason for doing this instead of receiving the config as a parameter? (which would mean editing the index file for best-practises, but it's what we are currently doing for all rules).

File: lib/rules/best-practises/no-empty-blocks.js [hunk 5]
```diff
@@ -50,27 +75,37 @@ class NoEmptyBlocksChecker extends BaseChecker {
...
     }
-    this._validateChildrenCount(node, 'statements')
-  }
 
-  StructDefinition(node) {
-    this._validateChildrenCount(node, 'members')
-  }
+    if (node.parent.type === 'ModifierDefinition') {
+      if (this.allowEmptyModifiers) {
+        if (!this._hasUnderscore(node)) {
```
Comment (juanpcapurro) [inline lib/rules/best-practises/no-empty-blocks.js:90]: I believe we can simply not handle modifiers without `_;` as they are forbidden by the compiler

File: lib/rules/best-practises/no-empty-blocks.js [hunk 6]
```diff
@@ -50,27 +75,37 @@ class NoEmptyBlocksChecker extends BaseChecker {
...
+        }
+      } else {
+        this._error(node)
+      }
+      return
+    }
 
-  EnumDefinition(node) {
-    this._validateChildrenCount(node, 'members')
+    if (node.parent.type === 'CatchClause') {
```
Comment (juanpcapurro) [inline lib/rules/best-practises/no-empty-blocks.js:99]: when are we checking the block is actually empty in this branch? am I missing something?

File: test/rules/best-practises/no-empty-blocks.js [hunk 7]
```diff
@@ -123,4 +123,99 @@ describe('Linter - no-empty-blocks', () => {
...
+        rules: {
+          'no-empty-blocks': ['warn', { allowEmptyModifiers: false }],
+        },
+      })
+
+      assertWarnsCount(report, 1)
+      assertErrorMessage(report, 'Code contains empty blocks')
+    })
+
+    it('should raise warning for completely empty modifier', () => {
```
Comment (juanpcapurro) [inline test/rules/best-practises/no-empty-blocks.js:145]: this I believe is a nice to have, but if implementation ends up being simpler by ignoring it, we can live with it

File: lib/rules/best-practises/no-empty-blocks.js [hunk 8]
```diff
@@ -16,32 +20,53 @@ const meta = {
...
         { description: 'empty receive function', code: 'receive () external {}' },
         { description: 'empty fallback function', code: 'fallback () external {}' },
         {
-          description: 'empty constructor with member initialization list',
-          code: 'constructor(uint param) Foo(param) Bar(param*2) {}',
+          description: 'empty constructor with parent init',
+          code: 'constructor(uint x) Base(x) {}',
         },
+        { description: 'empty modifier with _ allowed', code: 'modifier onlyOwner { _; }' },
+        { description: 'empty try-catch block allowed', code: 'try foo() {} catch {}' },
```
Comment (juanpcapurro) [inline lib/rules/best-practises/no-empty-blocks.js:36]: given the current rule implementation, this would cause an error on the empty  `try`

File: test/rules/best-practises/no-empty-blocks.js [hunk 9]
```diff
@@ -123,4 +123,99 @@ describe('Linter - no-empty-blocks', () => {
...
+
+    it('should not raise warning for empty modifier with _ when allowEmptyModifiers is true', () => {
+      const code = modifierWith('modifier onlyOwner { _; }')
+      const report = linter.processStr(code, {
+        rules: {
+          'no-empty-blocks': ['warn', { allowEmptyModifiers: true }],
+        },
+      })
+
+      assertNoWarnings(report)
```
Comment (juanpcapurro) [inline test/rules/best-practises/no-empty-blocks.js:194]: this report does not contain warnings, but does contain a parsing error since `code` evaluates to 

```
      contract A {

        modifier b() {
          modifier onlyOwner { _; }
        }

      }
```

**Review by juanpcapurro — CHANGES_REQUESTED** [review]
thanks for taking care of this!

There are a few items left
- it'd be nice to have an `allowEmptyTry` option, similar to `allowEmptyCatch`, as users might be okay with one, both, or none of them, depending on their particular codebase.
- add tests for cases where the rules should _not_ throw with the new setting set to `false` due to the modifier having some content besides `_;`
- update CHANGELOG.md referencing this PR

File: lib/rules/best-practises/no-empty-blocks.js (hunk 4 above)
Comment (juanpcapurro) [reply lib/rules/best-practises/no-empty-blocks.js]: you have to pass the config in as a parameter in `lib/rules/best-practises/index.js` for that to work

File: test/rules/best-practises/no-empty-blocks.js [hunk 10]
```diff
@@ -1,27 +1,75 @@
...
-  EMPTY_BLOCKS.forEach((curData) =>
-    it(`should raise warn for empty blocks ${label(curData)}`, () => {
-      const report = linter.processStr(curData, {
-        rules: { 'no-empty-blocks': 'warn' },
+    EMPTY_BLOCKS.forEach((curData) =>
+      it(`should raise warn for empty block: ${label(curData)}`, () => {
+        const report = linter.processStr(curData, {
+          rules: { 'no-empty-blocks': 'warn' },
+        })
+        assertWarnsCount(report, curData.includes('try') ? 2 : 1)
```
Comment (juanpcapurro) [inline test/rules/best-practises/no-empty-blocks.js:30]: I'd rather define a separate testcase for the `try` statement since it's the only on with two empty blocks, and would like to keep test logic as simple as possible

File: test/rules/best-practises/no-empty-blocks.js [hunk 11]
```diff
@@ -1,27 +1,75 @@
...
+      'contract Token { string private name; }',
+      funcWith('assembly { let x := 1 }'),
+    ]
+
+    VALID_BLOCKS.forEach((curData) =>
+      it(`should not raise warn for valid block: ${label(curData)}`, () => {
+        const report = linter.processStr(curData, {
+          rules: { 'no-empty-blocks': 'warn' },
+        })
+        assertNoWarnings(report)
```
Comment (juanpcapurro) [inline test/rules/best-practises/no-empty-blocks.js:55]: since we already had the false negative caused by a syntax error, I'd say we add an assertion on the report not including any errors either (here and for all other no-warning testcases in this suite). We should also consider editing the `assertNoWarnings` function to check both errors and warnings, if it's not also thoroughly used to check no warnings are emitted in the presence of errors

File: test/rules/best-practises/no-empty-blocks.js [hunk 12]
```diff
@@ -123,4 +171,230 @@ describe('Linter - no-empty-blocks', () => {
...
+  describe('Modifier cases', () => {
+    it('should raise warning for empty modifier', () => {
+      const code = contractWith('modifier onlyOwner() { }')
+      const report = linter.processStr(code, {
+        rules: { 'no-empty-blocks': 'warn' },
+      })
+      assertWarnsCount(report, 1)
+    })
+
+    it('should not raise warning for modifier with placeholder', () => {
```
Comment (juanpcapurro) [inline test/rules/best-practises/no-empty-blocks.js:243]: we should report on this unless the `allowEmptyModifiers` option is set to `true`. This is the normal 'empty modifiers' case, the one above is forbidden by the compiler. I encourage you to init a small foundry project and try this kind of things out, might make your life a lot easier :grin:

File: test/rules/best-practises/no-empty-blocks.js [hunk 13]
```diff
@@ -123,4 +171,230 @@ describe('Linter - no-empty-blocks', () => {
...
+      const report = linter.processStr(code, {
+        rules: { 'no-empty-blocks': 'warn' },
+      })
+      assertWarnsCount(report, 1)
+    })
+  })
+
+  describe('No empty blocks with configuration options', () => {
+    describe('Empty modifiers', () => {
+      const modifierCode = contractWith('modifier onlyOwner() { }')
```
Comment (juanpcapurro) [inline test/rules/best-practises/no-empty-blocks.js:315]: this should include the `_;` since it's required by the compiler

File: CHANGELOG.md [hunk 14]
```diff
@@ -1,4 +1,8 @@
 ## [4.0.1] - XXXX-XX-XX
+### Added
+- Added `allowEmptyTry` option to `no-empty-blocks` rule to allow empty try blocks
```
Comment (juanpcapurro) [inline CHANGELOG.md:3]: make it one item, showing the link to the PR as it is the case for other entries in the changelog, and the three options as sub-items

**Review by juanpcapurro — CHANGES_REQUESTED** [review]
blockers:
- [ ] semantic for what constitutes an empty modifier
- [ ] ensure we are not letting parsing errors mask false negatives for this rule

also, would it be good semver to merge this and publish it as part of 4.0.1? or should we delay merging it or start working on a separate branch for 4.1.0?

**Review by jahabeebs — COMMENTED** [review]
Hi Juan,

Thanks a lot for your feedback. I made the following changes in response to your PR comments

here is the relevant commit: https://github.com/solhint-community/solhint-community/pull/162/commits/693c49dc3f47de87cc1fd890d87dca648005f216

1) Cleaned up tests more
- Removed/modified several examples in the tests that did not parse correctly (both assembly & solidity examples)
2) Changelog
- Consolidated the Changelog entry to one item w/several sub-items & added the PR link, marked for 4.1.0
3) Rule refactoring
- Refactored the rule to use optional chaining when possible, simplified the code a bit, & added more comments

File: test/rules/best-practises/no-empty-blocks.js (hunk 11 above)
Comment (juanpcapurro) [reply test/rules/best-practises/no-empty-blocks.js:64]: seems reasonable. Did you create the issue?

File: lib/rules/best-practises/no-empty-blocks.js [hunk 15]
```diff
@@ -50,18 +89,47 @@ class NoEmptyBlocksChecker extends BaseChecker {
...
       return
     }
 
-    const isFallbackFunctionBlock = isFallbackFunction(node.parent)
-    const isReceiveFunctionBlock = isReceiveFunction(node.parent)
-    if (isFallbackFunctionBlock || isReceiveFunctionBlock) {
-      // ignore empty blocks in fallback or receive functions
+    // empty try block
+    if (node.parent?.type === 'TryStatement') {
+      if (node.parent.body === node && this.allowEmptyTry) return
```
Comment (juanpcapurro) [inline lib/rules/best-practises/no-empty-blocks.js:99]: in which case would `node.parent.body !== node` ? could the block be another attribute? in that case, would it make more sense to assert on the type of the parent?

Also if I remove that comparison all tests continue to pass

File: test/rules/best-practises/no-empty-blocks.js [hunk 16]
```diff
@@ -1,37 +1,94 @@
...
     contractWith('function () public payable { make1(); }'),
-    contractWith('receive() external payable {}'),
-    contractWith('fallback() external payable {}'),
+    contractWith('receive() external payable { revert(); }'),
+    contractWith('fallback() external payable { revert(); }'),
     funcWith('if (a < b) { make1(); }'),
     contractWith('struct Abc { uint a; }'),
     contractWith('enum Abc { Test1 }'),
     'contract A { uint private a; }',
-    funcWith('assembly { "literal" }'),
```
Comment (juanpcapurro) [inline test/rules/best-practises/no-empty-blocks.js:34]: why did you remove this? is it syntactically invalid?

File: test/rules/best-practises/no-empty-blocks.js [hunk 17]
```diff
@@ -123,4 +183,235 @@ describe('Linter - no-empty-blocks', () => {
...
+      funcWith('try foo() returns (uint) { } catch Error(string memory) { revert(); }'),
+      funcWith('try foo() {  } catch {  }'),
+    ]
+
+    TRY_CATCH_CASES.forEach((code) => {
+      it(`should detect empty blocks in try-catch: ${label(code)}`, () => {
+        const report = linter.processStr(code, {
+          rules: { 'no-empty-blocks': 'warn' },
+        })
+        const expectedWarnings = code.includes('{  } catch {  }') ? 2 : 1
```
Comment (juanpcapurro) [inline test/rules/best-practises/no-empty-blocks.js:228]: not in favor of having multi-branch testcases, I'd rather have two separate describes with some repeated code

File: test/rules/best-practises/no-empty-blocks.js [hunk 18]
```diff
@@ -123,4 +183,235 @@ describe('Linter - no-empty-blocks', () => {
...
+        assertWarnsCount(report, expectedWarnings)
+      })
+    })
+  })
+
+  describe('Constructor inheritance', () => {
+    it('should not raise warning for empty constructor with multiple inheritance', () => {
+      const code = `
+        contract C is A, B {
```
Comment (juanpcapurro) [inline test/rules/best-practises/no-empty-blocks.js:237]: if having a contract equivalent to this one but written like this:
```
contract C is A(1) B(2) {}
```
is considered valid by the compiler then we should probably add an exception (like this one, not configurable is okay)

File: test/rules/best-practises/no-empty-blocks.js [hunk 19]
```diff
@@ -123,4 +183,235 @@ describe('Linter - no-empty-blocks', () => {
...
+          require(value < 200);
+        }
+      `)
+      const report = linter.processStr(code, {
+        rules: { 'no-empty-blocks': 'warn' },
+      })
+      assertNoWarnings(report)
+    })
+
+    it('should not raise warning for modifier with multiple placeholders', () => {
```
Comment (juanpcapurro) [inline test/rules/best-practises/no-empty-blocks.js:281]: what if the modifier _only_ has multiple placeholders?

**Review by juanpcapurro — CHANGES_REQUESTED** [review]
have some questions on edge cases & potentially redundant code, but looks good otherwise
